File: Singleton/Singleton.cpp
#include <iostream>
using namespace std;

class Singleton
{
private:
    // Here will be the instance stored.
    static Singleton* m_instance;
    
    // Private constructor to prevent instancing.
    Singleton();
    
public:
    // Static access method.
    static Singleton* getInstance();
};

// Start with 0 to create a new object
Singleton* Singleton::m_instance = 0;

Singleton* Singleton::getInstance()
{
    if (m_instance == 0)
    {
        // if it's multi-thread symstem, use synchronized to prevent
        cout << "new Singleton" << endl;
        m_instance = new Singleton();
    }
    else
        cout << "already exist" << endl;
    
    return m_instance;
}

Singleton::Singleton()
{}

int main()
{
    //new Singleton(); // Don't work
    Singleton* s = Singleton::getInstance(); // Ok
    cout << s << endl;
    Singleton* r = Singleton::getInstance();
    // The Addresses will be the same.
    cout << r << endl;
    
}

File: Singleton/Singleton.lua
-- Inheritance Func
function FuncNew( obj )
	function obj:new( o )
		o = o or {}
	  	setmetatable( o, self )
	  	self.__index = self
	  	return o
	end
	return obj
end

------------------------------------------------------

Singleton = {}
function Singleton:Singleton()
	-- Here will be the instance stored.
	local m_instance = nil -- Private

	-- Static access method.
	function self:getInstance()
		if not m_instance then
			-- if it's multi-thread symstem, use synchronized to prevent
			print("new Singleton")
			m_instance = FuncNew( Singleton ):new()
		else
			print("already exist")
		end
		return m_instance
	end
end
-- Private constructor to prevent instancing.
Singleton:Singleton()
function Singleton:Singleton() end

------------------------------------------------------

-- local s = Singleton:new()
-- s:getInstance() -- Lua error
local s = Singleton:getInstance()
print(s)
-- Singleton:Singleton() -- Don't work
local r = Singleton:getInstance()
-- The addresses will be the same.
print(r)
